File: usedesk_sdk/src/main/java/ru/usedesk/sdk/ui/knowledgebase/DataViewModel.java
package ru.usedesk.sdk.ui.knowledgebase;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import io.reactivex.Single;
import io.reactivex.disposables.Disposable;
import ru.usedesk.sdk.utils.LogUtils;

public class DataViewModel<T> extends ViewModel {

    private static final String TAG = DataViewModel.class.getSimpleName();

    private MutableLiveData<DataOrMessage<T>> liveData = new MutableLiveData<>();
    private Disposable disposable;

    protected DataViewModel() {
        setData(new DataOrMessage<>(DataOrMessage.Message.LOADING));
    }

    public LiveData<DataOrMessage<T>> getLiveData() {
        return liveData;
    }

    protected void loadData(Single<T> single) {
        disposable = single.subscribe(this::onData, this::onThrowable);
    }

    @Override
    protected void onCleared() {
        super.onCleared();

        disposable.dispose();
    }

    protected void setData(DataOrMessage<T> DataOrMessage) {
        liveData.setValue(DataOrMessage);
    }

    private void onData(T data) {
        setData(new DataOrMessage<>(data));
    }

    private void onThrowable(Throwable throwable) {
        LogUtils.LOGE(TAG, throwable);

        setData(new DataOrMessage<>(DataOrMessage.Message.ERROR));
    }
}

File: knowledgebase-gui/src/main/java/ru/usedesk/knowledgebase_gui/common/DataViewModel.kt
package ru.usedesk.knowledgebase_gui.common

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import io.reactivex.Single
import io.reactivex.disposables.Disposable
import ru.usedesk.knowledgebase_gui.entity.DataOrMessage

open class DataViewModel<T> protected constructor() : ViewModel() {

    val liveData = MutableLiveData<DataOrMessage<T>>()
    private var disposable: Disposable? = null

    init {
        setData(DataOrMessage(DataOrMessage.Message.LOADING))
    }

    fun getLiveData(): LiveData<DataOrMessage<T>> {
        return liveData
    }

    protected fun loadData(single: Single<T>) {
        disposable = single.subscribe({
            onData(it)
        }) {
            setData(DataOrMessage(DataOrMessage.Message.ERROR))
        }
    }

    override fun onCleared() {
        super.onCleared()
        if (disposable?.isDisposed != true) {
            disposable?.dispose()
        }
    }

    private fun setData(DataOrMessage: DataOrMessage<T>) {
        liveData.value = DataOrMessage
    }

    protected open fun onData(data: T) {
        setData(DataOrMessage(data))
    }
}